### PR TITLE
Parse catch multiple exceptions

### PR DESCRIPTION
**Pull request by runz0rd**
For #103

Theres still the issue of detecting class names that do not implement Throwable. Any idea where can this fit in?

### CODE REVIEW

File: src/Parser.php [hunk 1]
```diff
@@ -2453,10 +2457,10 @@ private function parsePropertyDeclaration($parentNode, $modifiers) {
         return $propertyDeclaration;
     }
 
-    private function parseQualifiedNameList($parentNode) {
+    private function parseQualifiedNameList($parentNode, $delimiter = TokenKind::CommaToken) {
         return $this->parseDelimitedList(
             DelimitedList\QualifiedNameList::class,
-            TokenKind::CommaToken,
+            $delimiter,
```
Comment (mousetraps) [inline src/Parser.php]: Oh, I see what you meant in the delimited list PR - yeah, I don't think this is a super common case to handle, so I think it makes sense to special case this usage (or come up with some other idea that doesn't add additional memory usage.)

File: src/DiagnosticsProvider.php [hunk 2]
```diff
@@ -79,6 +79,17 @@ public static function getDiagnostics($node) : \Generator {
                     }
                 }
             }
+            elseif ($node instanceof Node\CatchClause) {
+                $lastElement = \end($node->qualifiedNameList->children);
+                if($lastElement instanceof Token && $lastElement->kind == TokenKind::BarToken) {
```
Comment (mousetraps) [inline src/DiagnosticsProvider.php]: nit: space between if and paren here and everywhere else. sorry, we don't have automatic code style checking yet #83

File: src/Parser.php [hunk 3]
```diff
@@ -1863,14 +1863,18 @@ private function parseCatchClause($parentNode) {
         $catchClause->parent = $parentNode;
         $catchClause->catch = $this->eat(TokenKind::CatchKeyword);
         $catchClause->openParen = $this->eat(TokenKind::OpenParenToken);
-        $catchClause->qualifiedName = $this->parseQualifiedName($catchClause); // TODO generate missing token or error if null
+        $catchClause->qualifiedNameList = $this->parseQualifiedNameList($catchClause, TokenKind::BarToken);
+        if(\is_null($catchClause->qualifiedNameList)) {
+            $catchClause->qualifiedNameList = new MissingToken(TokenKind::QualifiedName, $this->getCurrentToken()->fullStart);
+        }
```
Comment (mousetraps) [inline src/Parser.php]: Nice!

File: src/DiagnosticsProvider.php (hunk 2 above)
Comment (mousetraps) [inline src/DiagnosticsProvider.php]: Also === instead of ==

File: src/Parser.php [hunk 4]
```diff
@@ -1863,14 +1863,18 @@ private function parseCatchClause($parentNode) {
         $catchClause->parent = $parentNode;
         $catchClause->catch = $this->eat(TokenKind::CatchKeyword);
         $catchClause->openParen = $this->eat(TokenKind::OpenParenToken);
-        $catchClause->qualifiedName = $this->parseQualifiedName($catchClause); // TODO generate missing token or error if null
+        $catchClause->qualifiedNameList = $this->parseQualifiedNameList($catchClause, TokenKind::BarToken);
```
Comment (mousetraps) [inline src/Parser.php]: Let's also add a new test case to handle this scenario.

**Comment by mousetraps** [comment]
> Theres still the issue of detecting class names that do not implement Throwable. Any idea where can this fit in?

Cool idea, and probably best left as an exercise to the consumer :wink:. For that, we'd need to resolve imports, build a symbol table, etc., which are more "language service-y" things to do. cc @felixfbecker

File: src/Parser.php [hunk 5]
```diff
@@ -1863,14 +1863,18 @@ private function parseCatchClause($parentNode) {
...
+        if(\is_null($catchClause->qualifiedNameList)) {
```
Comment (felixfbecker) [inline src/Parser.php]: using `=== null` is a lot faster and does the same

**Comment by felixfbecker** [comment]
I agree that checking the class extends `Throwable` is the domain of a language service :)

File: tests/cases/php-langspec/exception_handling/odds_and_ends.php [hunk 6]
```diff
@@ -20,14 +20,22 @@ class X {}	// not derived from Exception
...
 	echo "L0: In catch-block int\n";
 }
+
+//catch (InvalidArgumentException | $e) { // Diagnosed; good
+//    echo "L0: In catch-block multiple Exception\n";
+//}
+
+catch (BadMethodCallException | InvalidArgumentException $e) {
+    echo "L0: In catch-block multiple Exception\n";
+}
```
Comment (runz0rd) [inline tests/cases/php-langspec/exception_handling/odds_and_ends.php]: Some of the grammar tests from this file were failing. This was intentionally added to test if a valid case would pass. Want this one removed?

Comment (mousetraps) [inline tests/cases/php-langspec/exception_handling/odds_and_ends.php]: Yep, let's revert this file but just make sure that there are handwritten test cases added for this scenario (which it looks like you already added). The spec tests are generated using [this script](https://github.com/Microsoft/tolerant-php-parser/blob/master/tests/generate_spec_tests.php) and will be overwritten whenever we take updates from [php-langspec](github.com/php/php-langspec). Original source [here](https://github.com/Microsoft/tolerant-php-parser/blob/master/php-langspec/tests/exception_handling/odds_and_ends.phpt).

**Comment by roblourens** [comment]
Hey @runz0rd, this looks ok to me - sorry it fell through the cracks. If you're still up for it, I would accept it if you can get it merged with the latest stuff in master.

Also, where did the `--TEST--` and `--EXPECT--` come from in the test?

**Comment by mousetraps** [comment]
> Also, where did the --TEST-- and --EXPECT-- come from in the test? I don't see that in other php-langspec tests. CI is failing, not sure whether that's the reason.

@roblourens those look like the raw php language spec test format. The files in `tests/cases/php-langspec` are autogenerated using the `tests/generate_spec_tests.php` script, and shouldn't be changed manually.

If upstream changes happen in php/php-langspec, they can be pulled in using git subtree. Then run `generate_spec_tests.php` to regenerate test cases for the parser.

**Comment by roblourens** [comment]
Got it, thanks!

**Comment by roblourens** [comment]
This will be handled in #244, and now would be a breaking change.

**Pull request closed without merging**